### ISSUE TITLE
Version 2 roadmap + questions and ideas

### ISSUE THREAD

**Issue by stojanovic** [stojanovic/scottyjs#18]
If you are using Scotty.js I would like to know what do you like and what should be improved. Feel free to comment bellow.

We started working on version 2 that will have breaking changes. This is a description of what can you expect in v2.

# Scotty.js, version 2

Scotty v1 was a MVP, our goal was to see if it can be useful to us and anyone else. Now that it has some traction, it's time to fix it up :)

## Version 2 features

Here's the initial version 2 features that we will work on. This list is not final. Feel free to give us feedback on v1 and tell us what would you like to see in v2.

1. Running Scotty command first time will create `scotty.json` file with all the options. This file will allow us to share Scotty projects, ie. if you commit this file with your repository, anyone with AWS permissions can simply run the `scotty` command to deploy the app.
2. `create`, `update` and `destroy` commands will be added. Running the `scotty` command will run `scotty create` or `scotty update` depending on `scotty.json` file existence.
3. Proper destroy command will be available. It will destroy the S3 bucket and CloudFront distribution if they were created by Scotty.
4. Custom domains feature. We have PR for that right now, but it's not finished, version 2 should have that integrated.
5. Clearing CloudFront cache on the update.
6. Support for multiple versions (ie. dev, staging and production).
7. TBA

**Comment by duro** [comment]
Rollbacks!

Take a look at the way ember-deploy works: http://ember-cli-deploy.com/docs/v1.0.x/s3-walkthrough/

This is relatively trivial for an app that uses Create React App with Webpack as a bundler, and the compiled assets are already written with file names that are hashed.

All that would be required would be to hash the index filename, much like ember-cli-deploy does, and then on a fresh deploy/update you would copy this latest hashed filename to the index.html file.

Add a CLI command to `scotty ls` which returns all the hashed versions for reference, and then a `scotty rollback <hash>` command that simply copies the hashed index file on top of the existing index.html.

I'm just riffing here. Truthfully, I've never used scottyjs, and simply stumbled up on this repo as I was searching for something I could use with react that worked like ember-cli-deploy (before I decided to write my own).

**Comment by stojanovic** [comment]
@duro Interesting. It's easier to do so for SPA then a static website. But I'll see if we can add that. Thanks

**Comment by amirilovic** [comment]
Support for SSL certificate for custom domains. This is something that should be done as part of or after actual custom domain support. Custom domain support should be split into smaller features: CloudFront configuration, SSL support, route 53 configuration.

**Comment by stojanovic** [comment]
@amirilovic You are totally right :) I should add better explanation for the tasks, for now I'll add SSL support in the list above.

**Comment by alekseykulikov** [comment]
Lambda@Edge support! This allows to configure custom headers, custom routing and more.

Possible config prototype:
```json
{
  "name": "alekseykulikov.com",
  "routes": {
    "/": "index.html",
    "/hire": "hire.html",
    "*": "404.html"
  },
  "extra-headers": {
    "Strict-Transport-Security": "max-age=31536000; includeSubdomains; preload",
    "X-Content-Type-Options": "nosniff",
    "X-Frame-Options": "DENY",
    "X-XSS-Protection": "1; mode=block"
  },
  "redirects": {
    "/github": "https://github.com/alekseykulikov"
  }
}
```

**Comment by luanmuniz** [comment]
Use custom credentials. I have about 5 different aws credentials, we should be able to pass a profile name.

**Comment by stojanovic** [comment]
Good point, thanks for PR, I'll check it soon.
That was planned for next version but I forgot to add it on the list.

**Comment by stojanovic** [comment]
Version 2 is finally on its way. It's not done yet, but I am working on it.

![vykchjcbs9](https://user-images.githubusercontent.com/613478/35808643-bede9f48-0a86-11e8-90aa-8aff9ef572f3.gif)

**Comment by luanmuniz** [comment]
@stojanovic Anything i can help with to make v2 come faster?

**Comment by stojanovic** [comment]
@luanmuniz yes, that would be awesome, I stuck because of my day job and other things. I'll try to push everything and document what's left to be done tomorrow, and then you can jump in and help. I should have more time now.

**Comment by luanmuniz** [comment]
@stojanovic Don't need to go into details now. Just put the high-level tasks (Like To-do list) that are left and if needed we go into details later!

I use this project on the daily basis. I'm the one that needs to thank you! haha It's the least i can do

**Comment by rtwall** [comment]
An option to _Restrict bucket access_ would be nice.

CloudFront Origin Setting:
![image](https://user-images.githubusercontent.com/7053824/43011362-9b8689c4-8c19-11e8-90c1-7d9c7aa77bde.png)
